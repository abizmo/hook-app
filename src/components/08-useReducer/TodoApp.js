import React, { useEffect, useReducer } from 'react'

import TodoAddForm from './components/TodoAddForm'
import TodoList from './components/TodoList'
import todoReducer from './todoReducer'

import './styles.css'

const init = () => JSON.parse(localStorage.getItem('todos')) || []

const TodoApp = () => {
  const [todos, dispatch ] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos))
  }, [todos])

  const handleDelete = (id) => dispatch({
    type: 'DELETE_TODO',
    payload: id
  })

  const handleToggle = (id) => dispatch({
    type: 'TOGGLE_TODO',
    payload: id
  })

  const handleSubmit = (todo) => {
    const newTodo = {
      id: new Date().getTime(),
      todo,
      done: false,     
    }

    dispatch({
      type: 'ADD_TODO',
      payload: newTodo,
    })
  }

  return (
    <div className="container pt-3">
      <h3>TodoApp ({todos.length})</h3>
      <hr />
      <TodoAddForm onAddTodo={handleSubmit} />
      <TodoList
        todos={todos}
        onDeleteTodo={handleDelete}
        onToggleTodo={handleToggle}
      />
    </div>
  )
}

export default TodoApp
