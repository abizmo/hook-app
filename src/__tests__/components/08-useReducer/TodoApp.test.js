import { mount, shallow } from 'enzyme'
import { act } from '@testing-library/react'

import TodoApp from "../../../components/08-useReducer/TodoApp"
import demoTodos from '../../../fixtures/demoTodos'

describe('TodoApp tests', () => {
  const wrapper = shallow(<TodoApp />)

  Storage.prototype.setItem = jest.fn()

  test('should render and match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  
  test('should add a Todo', () => {
    const wrapper = mount(<TodoApp />)

    act( () => {
      wrapper.find('TodoAddForm').prop('onAddTodo')(demoTodos[0].todo)
      wrapper.find('TodoAddForm').prop('onAddTodo')(demoTodos[1].todo)
    })

    expect(wrapper.find('h3').text().trim()).toBe('TodoApp (2)')
    expect(localStorage.setItem).toHaveBeenCalledTimes(2)
  })

})
