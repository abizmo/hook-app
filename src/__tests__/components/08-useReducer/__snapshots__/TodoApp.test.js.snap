// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp tests should render and match snapshot 1`] = `
<div
  className="container pt-3"
>
  <h3>
    TodoApp (
    0
    )
  </h3>
  <hr />
  <TodoAddForm
    onAddTodo={[Function]}
  />
  <TodoList
    onDeleteTodo={[Function]}
    onToggleTodo={[Function]}
    todos={Array []}
  />
</div>
`;
